src/lib/comms/sol-socket-impl-zephyr.c

```diff
@@ -0,0 +1,383 @@
+    nano_sem_give(&socket->lock);
+
+    return &socket->base;
+
+socket_error:
+    errno = ENOMEM;
+    return NULL;
+
+unsupported_family:
+    errno = EAFNOSUPPORT;
+    return NULL;
+}
+
+static void
+sol_socket_zephyr_del(struct sol_socket *s)
+{
+    struct sol_socket_zephyr *socket = (struct sol_socket_zephyr *)s;
+    struct net_context *ctx;
+    struct net_buf *buf;
+    uint16_t idx;
+
+    sol_ptr_vector_del_element(&bound_sockets, socket);
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&socket->bufs, buf, idx)
+        ip_buf_unref(buf);
+    sol_ptr_vector_clear(&socket->bufs);
+
+    net_context_put(socket->unicast_context);
+    SOL_PTR_VECTOR_FOREACH_IDX (&socket->mcast_contexts, ctx, idx)
+        net_context_put(ctx);
+    sol_ptr_vector_clear(&socket->mcast_contexts);
+
+    free(socket);
+}
+
+static int
+sol_socket_zephyr_set_on_read(struct sol_socket *s, bool (*cb)(void *data, struct sol_socket *s), const void *data)
+{
+    struct sol_socket_zephyr *socket = (struct sol_socket_zephyr *)s;
+
+    socket->read.cb = cb;
+    socket->read.data = data;
+
+    return 0;
+}
+
+static bool
+write_timeout_cb(void *data)
+{
+    struct sol_socket_zephyr *socket = data;
+
+    if (socket->write.cb((void *)socket->write.data, &socket->base))
+        return true;
+
+    socket->write_timeout = NULL;
+    return false;
+}
+
+static int
+sol_socket_zephyr_set_on_write(struct sol_socket *s, bool (*cb)(void *data, struct sol_socket *s), const void *data)
+{
+    struct sol_socket_zephyr *socket = (struct sol_socket_zephyr *)s;
+
+    if (cb && !socket->write_timeout) {
+        socket->write_timeout = sol_timeout_add(0, write_timeout_cb, socket);
+        SOL_NULL_CHECK(socket->write_timeout, -ENOMEM);
+    } else if (!cb && socket->write_timeout) {
+        sol_timeout_del(socket->write_timeout);
+        socket->write_timeout = NULL;
+    }
+
+    socket->write.cb = cb;
+    socket->write.data = data;
+
+    return 0;
+}
+
+static ssize_t
+sol_socket_zephyr_recvmsg(struct sol_socket *s, void *buf, size_t len, struct sol_network_link_addr *cliaddr)
+{
+    struct sol_socket_zephyr *socket = (struct sol_socket_zephyr *)s;
+    struct net_buf *netbuf;
+    size_t buflen = -EAGAIN;
+
+    nano_task_sem_take(&socket->lock, TICKS_UNLIMITED);
+
+    netbuf = sol_ptr_vector_get(&socket->bufs, 0);
+    if (netbuf) {
+        buflen = ip_buf_appdatalen(netbuf);
+        if (buf)
+            sol_ptr_vector_del(&socket->bufs, 0);
+    }
+
+    nano_task_sem_give(&socket->lock);
+
+    if (!buf)
+        return buflen;
+
+    if (buflen > len)
+        buflen = len;
+
+    if (cliaddr) {
+        cliaddr->family = SOL_NETWORK_FAMILY_INET6;
+        cliaddr->port = uip_ntohs(NET_BUF_UDP(netbuf)->srcport);
+        memcpy(cliaddr->addr.in6, &NET_BUF_IP(netbuf)->srcipaddr, sizeof(cliaddr->addr.in6));
+    }
+
+    memcpy(buf, ip_buf_appdata(netbuf), buflen);
+
+    ip_buf_unref(netbuf);
+
+    return buflen;
+}
+
+#include <ip/simple-udp.h>
```

Comment:
On Mon, 28 Mar 2016 11:45:38 -0700, Ederson de Souza wrote:
> > +        buflen = len;
> > +
> > +    if (cliaddr) {
> > +        cliaddr->family = SOL_NETWORK_FAMILY_INET6;
> > +        cliaddr->port = uip_ntohs(NET_BUF_UDP(netbuf)->srcport);
> > +        memcpy(cliaddr->addr.in6, &NET_BUF_IP(netbuf)->srcipaddr, sizeof(cliaddr->addr.in6));
> > +    }
> > +
> > +    memcpy(buf, ip_buf_appdata(netbuf), buflen);
> > +
> > +    ip_buf_unref(netbuf);
> > +
> > +    return buflen;
> > +}
> > +
> > +#include <ip/simple-udp.h>
> 
> Is there any reason for this `include` to be here? (Compared to be on file header...)

Because it's a horrible hack, but unfortunately we need it to be able to
get the udp_conn pointer from the net_context.
If it were possible, I'd also have an #uninclude after the function that
uses it.